safehttp/plugins/csp/csp_test.go

```diff
@@ -230,3 +233,119 @@ func TestNonceEmptyContext(t *testing.T) {
 		t.Errorf("Nonce(ctx) got nonce: %v want: %v", n, want)
 	}
 }
+
+func TestCommitNonce(t *testing.T) {
+	rec := safehttptest.NewResponseRecorder()
+	req := safehttptest.NewRequest(safehttp.MethodGet, "https://foo.com/pizza", nil)
+	req.SetContext(context.WithValue(req.Context(), ctxKey{}, "pizza"))
+
+	it := Interceptor{}
+	tr := safehttp.TemplateResponse{FuncMap: map[string]interface{}{}}
+	it.Commit(rec.ResponseWriter, req, tr, nil)
+
+	nonce, ok := tr.FuncMap["CSPNonce"]
+	if !ok {
+		t.Error(`tr.FuncMap["CSPNonce"]: got nil, want nonce`)
+	}
+
+	fn, ok := nonce.(func() string)
+	if !ok {
+		t.Errorf(`tr.FuncMap["CSPNonce"]: got %T, want "func() string"`, fn)
+	}
+	if want, got := "pizza", fn(); want != got {
+		t.Errorf(`tr.FuncMap["CSPNonce"](): got %s, want %s`, got, want)
+	}
+
+	if want, got := safehttp.StatusOK, rec.Status(); want != got {
+		t.Errorf("rec.Status(): got %v, want %v", got, want)
+	}
+
+	if diff := cmp.Diff(map[string][]string{}, map[string][]string(rec.Header())); diff != "" {
+		t.Errorf("rec.Header() mismatch (-want +got):\n%s", diff)
+	}
+
+	if want, got := "", rec.Body(); got != want {
+		t.Errorf("rec.Body(): got %q want %q", got, want)
+	}
+}
+
+func TestCommitMissingNonce(t *testing.T) {
+	rec := safehttptest.NewResponseRecorder()
+	req := safehttptest.NewRequest(safehttp.MethodGet, "https://foo.com/pizza", nil)
+	req.SetContext(context.Background())
+
+	it := Interceptor{}
+	tr := safehttp.TemplateResponse{FuncMap: map[string]interface{}{}}
+	it.Commit(rec.ResponseWriter, req, tr, nil)
+
+	wantFuncMap := map[string]interface{}{}
+	if diff := cmp.Diff(wantFuncMap, tr.FuncMap); diff != "" {
+		t.Errorf("tr.FuncMap: mismatch (-want +got):\n%s", diff)
+	}
+
+	if want, got := safehttp.StatusInternalServerError, rec.Status(); got != want {
+		t.Errorf("rec.Status(): got %v, want %v", got, want)
+	}
+	wantHeaders := map[string][]string{
+		"Content-Type":           {"text/plain; charset=utf-8"},
+		"X-Content-Type-Options": {"nosniff"},
+	}
+	if diff := cmp.Diff(wantHeaders, map[string][]string(rec.Header())); diff != "" {
+		t.Errorf("rw.header mismatch (-want +got):\n%s", diff)
+	}
+	if want, got := "Internal Server Error\n", rec.Body(); got != want {
+		t.Errorf("rec.Body(): got %q want %q", got, want)
+	}
+}
+
+func TestCommitNotTemplateResponse(t *testing.T) {
+	rec := safehttptest.NewResponseRecorder()
+	req := safehttptest.NewRequest(safehttp.MethodGet, "https://foo.com/pizza", nil)
+
+	it := Interceptor{}
+	it.Commit(rec.ResponseWriter, req, safehttp.NoContentResponse{}, nil)
+
+	if want, got := safehttp.StatusOK, rec.Status(); want != got {
+		t.Errorf("rec.Status(): got %v, want %v", got, want)
+	}
+
+	if diff := cmp.Diff(map[string][]string{}, map[string][]string(rec.Header())); diff != "" {
+		t.Errorf("rec.Header() mismatch (-want +got):\n%s", diff)
+	}
+
+	if want, got := "", rec.Body(); got != want {
+		t.Errorf("rec.Body(): got %q want %q", got, want)
+	}
+
+}
+
+func TestBeforeCommit(t *testing.T) {
```

Comment:
What is this testing?

As a framework, we don't guarantee that e.g. the `ResponseWriter` or the request are going to stay the same between these calls.

Reply:
That is right, other interceptors or handlers could change the `ResponseWriter`, but I implemented this test to verify that the Before and Commit phases work together. Is there a better way to do it?

Reply:
Well, but this just tests that a nonce found it's way to the context. It just feels like a quite incomplete unit test.

Wouldn't it make more sense to instead create a test that actually runs through a handler? This way you can assert that:
- the headers are set with a given nonce
- the nonce is available inside the handler through `csp.Nonce()` call (this is important for migration, when a service wants to adopt the plugin to generate nonces and headers, but doesn't have automated nonce-injection implemented yet)
- the same nonce is passed in `FuncMap["CSPNonce"]` in the `TemplateResponse` and can be used by the dispatcher

Reply:
Similarily for the XSRF token.

Reply:
I see, yeah, that might be better. Which headers dym?

Reply:
CSP. :)

Reply:
In order to test that functionality we need a `ServeMux` on which to install the plugins so I think it's better if I extend the integration test.

Reply:
Fair. I'd remove this test then, as it doesn't really add much value if we have these unit tests for individual functions and we have an integration test for a real scenario (i.e. used with mux).